sdk/dynamicconfig/sdk/metric/sdk.go

```diff
@@ -0,0 +1,625 @@
+		// allocation.
+		sortSlice label.Sortable
+
+		// inst is a pointer to the corresponding instrument.
+		inst *syncInstrument
+
+		// current implements the actual RecordOne() API,
+		// depending on the type of aggregation.  If nil, the
+		// metric was disabled by the exporter.
+		current    export.Aggregator
+		checkpoint export.Aggregator
+	}
+
+	instrument struct {
+		meter      *Accumulator
+		descriptor metric.Descriptor
+	}
+
+	asyncInstrument struct {
+		instrument
+		// recorders maps ordered labels to the pair of
+		// labelset and recorder
+		recorders map[label.Distinct]*labeledRecorder
+	}
+
+	labeledRecorder struct {
+		observedEpoch int64
+		labels        *label.Set
+		observed      export.Aggregator
+	}
+)
+
+var (
+	_ api.MeterImpl     = &Accumulator{}
+	_ api.AsyncImpl     = &asyncInstrument{}
+	_ api.SyncImpl      = &syncInstrument{}
+	_ api.BoundSyncImpl = &record{}
+
+	ErrUninitializedInstrument = fmt.Errorf("use of an uninitialized instrument")
+)
+
+func (inst *instrument) Descriptor() api.Descriptor {
+	return inst.descriptor
+}
+
+func (a *asyncInstrument) Implementation() interface{} {
+	return a
+}
+
+func (s *syncInstrument) Implementation() interface{} {
+	return s
+}
+
+func (a *asyncInstrument) observe(number api.Number, labels *label.Set) {
+	if err := aggregator.RangeTest(number, &a.descriptor); err != nil {
+		global.Handle(err)
+		return
+	}
+	recorder := a.getRecorder(labels)
+	if recorder == nil {
+		// The instrument is disabled according to the
+		// AggregatorSelector.
+		return
+	}
+	if err := recorder.Update(context.Background(), number, &a.descriptor); err != nil {
+		global.Handle(err)
+		return
+	}
+}
+
+func (a *asyncInstrument) getRecorder(labels *label.Set) export.Aggregator {
+	lrec, ok := a.recorders[labels.Equivalent()]
+	if ok {
+		if lrec.observedEpoch == a.meter.currentEpoch {
+			// last value wins for Observers, so if we see the same labels
+			// in the current epoch, we replace the old recorder
+			a.meter.processor.AggregatorFor(&a.descriptor, &lrec.observed)
+		} else {
+			lrec.observedEpoch = a.meter.currentEpoch
+		}
+		a.recorders[labels.Equivalent()] = lrec
+		return lrec.observed
+	}
+	var rec export.Aggregator
+	a.meter.processor.AggregatorFor(&a.descriptor, &rec)
+	if a.recorders == nil {
+		a.recorders = make(map[label.Distinct]*labeledRecorder)
+	}
+	// This may store nil recorder in the map, thus disabling the
+	// asyncInstrument for the labelset for good. This is intentional,
+	// but will be revisited later.
+	a.recorders[labels.Equivalent()] = &labeledRecorder{
+		observed:      rec,
+		labels:        labels,
+		observedEpoch: a.meter.currentEpoch,
+	}
+	return rec
+}
+
+// acquireHandle gets or creates a `*record` corresponding to `kvs`,
+// the input labels.  The second argument `labels` is passed in to
+// support re-use of the orderedLabels computed by a previous
+// measurement in the same batch.   This performs two allocations
```

Comment:
Nit: remove one of the spaces in these three consecutive spaces

Reply:
Interesting situation: this was stuff I directly copied from the existing sdk codebase. I deleted it anyway.